[DOCS] Documents the rerank task type of the Inference API

szabosteve opened the pull request:
## Overview

This PR adds documentation for the `rerank` task type usable with the `cohere` service of the inference API (PUT and POST).

### Preview

* [PUT inference](https://elasticsearch_bk_107404.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/put-inference-api.html)
* [POST inference](https://elasticsearch_bk_107404.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/post-inference-api.html)

szabosteve committed:
[DOCS] Documents the rerank task type of the Inference API.

szabosteve committed:
[DOCS] Fixes JSON.

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

szabosteve committed:
[DOCS] Fixes section levels.

maxhniebergall commented on docs/reference/inference/put-inference.asciidoc:
```diff
@@ -57,24 +57,27 @@ The unique identifier of the {infer} endpoint.
 The type of service supported for the specified task type.
 Available services:
-* `cohere`: specify the `text_embedding` task type to use the Cohere service.
+* `cohere`: specify the `text_embedding` or the `rerank `task type to use the
```
```suggestion
* `cohere`: specify the `text_embedding` or the `rerank` task type to use the
```

maxhniebergall commented on docs/reference/inference/put-inference.asciidoc:
```diff
@@ -100,18 +103,25 @@ the same name and the updated API key.
-https://docs.cohere.com/reference/embed[Cohere docs]. Defaults to
-`embed-english-v2.0`.
+The name of the model to use for the {infer} task.
+Available values for `rerank`:
+* `rerank-english-v3.0`,
+* `rerank-multilingual-v3.0`,
+* `rerank-english-v2.0`,
+* `rerank-multilingual-v2.0`.
+
+To review the available `text_embedding` models, refer to the
+https://docs.cohere.com/reference/embed[Cohere docs]. The default value for
+`text_embedding` is `embed-english-v2.0`.
```
I think rather than listing all of the available models, it probably makes more sense to just mention that the available models are listed on the cohere website. Then we won't need to update our docs when they release new models.

maxhniebergall replied:
https://docs.cohere.com/reference/rerank-1

maxhniebergall commented on docs/reference/inference/put-inference.asciidoc:
```diff
@@ -210,11 +220,37 @@ allocations. Must be a power of 2. Max allowed value is 32.
+`return_documents`::
+(Optional, boolean)
+For `cohere` service only. Specifies whether to return results with the doc
+text. If `false`, the API returns a list of `index` and `relevance score` value
+where `index`` is inferred from the doc list passed into the request. If `ture`, 
+the API returns an ordered list of `index`, `text` and `relevance score` values
+where `index` and `text` refers to the doc list passed into the request.
```
I'm not sure that we need the sentences "If `false` ...", "If `true`...". I think that just the previous sentence could be clear enough. We could maybe update that sentence to say "Specify whether to return doc text within the results.".

maxhniebergall commented on docs/reference/inference/put-inference.asciidoc:
```diff
@@ -277,6 +306,27 @@ PUT _inference/text_embedding/cohere-embeddings
+
+[source,console]
+------------------------------------------------------------
+PUT _inference/rerank/cohere-rerank 
```
looks great!

maxhniebergall commented on docs/reference/inference/put-inference.asciidoc:
```diff
@@ -210,11 +220,37 @@ allocations. Must be a power of 2. Max allowed value is 32.
+used for abuse detection.
+=====
++
+.`task_settings` for the `rerank` task type
```
love it

maxhniebergall commented on docs/reference/inference/post-inference.asciidoc:
```diff
@@ -64,30 +65,43 @@ seconds.
 [NOTE]
 ====
-Inference endpoints for the `completion` task type currently only support a single string as input.
+Inference endpoints for the `completion` task type currently only support a
```
I think we should at Tim Grein to review the completion components here https://github.com/elastic/elasticsearch/pull/106476

szabosteve replied:
He was the one who originally added the docs here: https://github.com/elastic/elasticsearch/pull/106876
The only change I've done here is some change of the markup.

maxhniebergall commented on docs/reference/inference/post-inference.asciidoc:
```diff
@@ -99,39 +113,90 @@ The API returns the following response:
 
+[discrete]
+[[inference-example-rerank]]
+===== Rerank example
```
It might be useful to reference this document somewhere in our docs https://docs.cohere.com/docs/elasticsearch-and-cohere#build-a-hybrid-search-index-with-cohere-and-elasticsearch

szabosteve replied:
I've added the link in the latest commit.

maxhniebergall reviewed (changes requested):
Thanks Istvan!

szabosteve committed:
[DOCS] Addresses feedback.

szabosteve commented:
Thank you, Max. I've addressed your comments via [9e0dd94](https://github.com/elastic/elasticsearch/pull/107404/commits/9e0dd94dbbebb18de1ad7e623dacdd04c2facfd6).

szabosteve commented:
@elasticsearchmachine update branch

szabosteve committed:
Merge branch 'main' into rerank-cohere-inference

maxhniebergall reviewed (approved):
LGTM
